src/rs/config.rs

```diff
@@ -0,0 +1,537 @@
+        }
+    }
+
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn set_app_name(mut self, value: String) -> Self {
+        self.app_name = Some(CString::new(value.as_bytes()).unwrap());
+        self
+    }
+
+    pub fn set_execution_profile(mut self, value: ExecutionProfile) -> Self {
+        self.execution_profile = value;
+        self
+    }
+}
+
+/// Configures how to process a registration's workload.
+#[derive(Debug, PartialEq, Clone)]
+pub enum ExecutionProfile {
+    LowLatency,
+    MaxThroughput,
+    Scavenger,
+    RealTime,
+}
+
+impl Default for ExecutionProfile {
+    fn default() -> Self {
+        Self::LowLatency
+    }
+}
+
+impl From<ExecutionProfile> for crate::ffi::QUIC_EXECUTION_PROFILE {
+    fn from(value: ExecutionProfile) -> Self {
+        match value {
+            ExecutionProfile::LowLatency => {
+                crate::ffi::QUIC_EXECUTION_PROFILE_QUIC_EXECUTION_PROFILE_LOW_LATENCY
+            }
+            ExecutionProfile::MaxThroughput => {
+                crate::ffi::QUIC_EXECUTION_PROFILE_QUIC_EXECUTION_PROFILE_TYPE_MAX_THROUGHPUT
+            }
+            ExecutionProfile::Scavenger => {
+                crate::ffi::QUIC_EXECUTION_PROFILE_QUIC_EXECUTION_PROFILE_TYPE_SCAVENGER
+            }
+            ExecutionProfile::RealTime => {
+                crate::ffi::QUIC_EXECUTION_PROFILE_QUIC_EXECUTION_PROFILE_TYPE_REAL_TIME
+            }
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct CredentialConfig {
+    credential_flags: CredentialFlags,
+    credential: Credential,
+    principle: Option<CString>, // TODO: support async handler.
+    allowed_cipher_suites: AllowedCipherSuiteFlags,
+    ca_certificate_file: Option<CString>,
+}
+
+impl CredentialConfig {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// flags are additive when called multiple times.
+    pub fn set_credential_flags(mut self, value: CredentialFlags) -> Self {
+        self.credential_flags |= value;
+        self
+    }
+
+    pub fn set_credential(mut self, value: Credential) -> Self {
+        self.credential = value;
+        self
+    }
+
+    pub fn set_principle(mut self, value: String) -> Self {
+        self.principle = Some(CString::new(value.as_bytes()).unwrap());
+        self
+    }
+
+    pub fn set_allowed_cipher_suites(mut self, value: AllowedCipherSuiteFlags) -> Self {
+        self.credential_flags |= CredentialFlags::SET_ALLOWED_CIPHER_SUITES;
+        self.allowed_cipher_suites = value;
+        self
+    }
+
+    pub fn set_ca_certificate_file(mut self, value: String) -> Self {
+        self.credential_flags |= CredentialFlags::SET_CA_CERTIFICATE_FILE;
+        self.ca_certificate_file = Some(CString::new(value.as_bytes()).unwrap());
+        self
+    }
+
+    /// # Safety
+    /// ffi type returned needs to have the lifetime of self.
+    pub unsafe fn as_ffi(&self) -> QUIC_CREDENTIAL_CONFIG {
```

Comment:
Just wondering, is there a reason this one is implemented as unsafe instead of following the pattern of a `raw` member like other types?

Reply:
Ideally as_ffi() conversion should be implemented for all wrapper types and wrapper types should not contain the ffi types as fields. The reasons is that ffi type field makes the wrapper type unmovable, due to ptr in ffi points to some memory in the wrapper struct, and if user moves the wrapper type to another scope or return it from function, u get ptr pointing to wrong location.
The places where ffi type are fields is due to the inner ffi type memory needs to be valid after QUIC_CREDENTIAL_CONFIG is constructed. If there are internal ffi ptr they point to heap memory so the above-mentioned problem does not occur.

Anyway, I made all ffi functions crate level, so that users cannot call them, and we have the leeway to change them in future.

Reply:
I see, I didn't realize the move issue.

For the types using `ffi_ref`, Isn't that replacing one issue (the Rust type must outlive the ffi type) with another issue (the Rust type must not be moved) then? I don't see anything preventing the user from moving the Rust type and invalidating the ffi object.

Maybe populating the `raw` member only when `as_ffi_ref` is called (that way, the struct can be moved until the ref is requested) would help?
I imagine there is a combination of wrappers and PhantomData that would express that fully.

In any case, it can be done later. Maybe a comment would help not to forget the types must not be moved though.

Reply:
If the ffi_ref or as_ffi_ref  returned cannot be moved, the function is marked as unsafe, and user needs to take care of. Some of the ffi_ref is movable, and thus they are safe.
Changing raw member when as_ffi_ref is called requires mutability change on the function signature.

We can standardize more in future PRs.

Reply:
Ah, got it, I missed one level of indirection. Thanks for taking the time to explain.